src/nsrt_learning/side_predicate_learning.py

```diff
@@ -409,53 +409,24 @@ def _sideline(self) -> List[PartialNSRTAndDatastore]:
-                    # Before we can lift missing_effects, we need to add new
-                    # entries to obj_to_var to account for the situation where
-                    # missing_effects contains objects that were not in
-                    # the ground operator's parameters.
-                    var_names = [var.name for var in obj_to_var.values()]
-                    assert all(name.startswith("?x") for name in var_names)
-                    max_var_name = max(var_names,
-                                       key=lambda name: int(name[2:]),
-                                       default="?x-1")
-                    max_var_num = int(max_var_name[2:])
-                    for effect in missing_effects:
-                        for obj in effect.objects:
-                            if obj not in obj_to_var:
-                                max_var_num += 1
-                                new_var = Variable("?x" + str(max_var_num),
-                                                   obj.type)
-                                obj_to_var[obj] = new_var
-                    # Finally, we can lift missing_effects and update the PNAD.
-                    updated_params = sorted(obj_to_var.values())
-                    updated_add_effects = pnad.op.add_effects | {
-                        a.lift(obj_to_var)
-                        for a in missing_effects
-                    }
-                    self._update_pnad_with_params_and_add_effects(
-                        pnad, updated_params, updated_add_effects)
+                    # We now try to make the PNAD's add effects more specific.
+                    # Since we have started with the most general operators,
+                    # and we are making them more specific, it is guaranteed
+                    # that such a grounding exists, otherwise our 1:1
+                    # option:operator assumption is violated.
+                    new_pnad = self._try_specifizing_pnad(necessary_add_effects,
+                                                          pnad, segment)
+                    assert new_pnad is not None, "1:1 assumption violated"
+                    assert new_pnad.option_spec == pnad.option_spec
+                    pnad = new_pnad
+                    del new_pnad  # unused from here
+                    param_opt_to_pnad[option.parent] = pnad
                     # After all this, the unification call that failed earlier
-                    # (leading us to enter this if statement) should now work.
-                    unifies, _ = utils.find_substitution(
+                    # (leading us into the current if statement) should work.
+                    unifies, obj_to_var_ent = utils.find_substitution(
                         pnad.op.add_effects | opt_vars_atom,
                         necessary_add_effects | opt_objs_atom)
+                    obj_to_var = cast(Dict[Object, Variable], obj_to_var_ent)
```

Comment:
i know that this pattern is used elsewhere, but don't we have an `ObjToVarSub` struct that we can use here?

Reply:
thx, it wasn't in too many places, so i fixed them all